design/baremetal-operator/bmh_non-bootable_iso.md

```diff
@@ -0,0 +1,153 @@
+which in this proposal is Redfish.
+
+## Design Details
+
+Create a new Custom Resource(CR), `DataImage`, with spec field `Url` and the same `Name`
+and `Namespace` as the `BareMetalHost` where we are attaching the image.
+
+To attach an ISO, first the `DataImage` CR object is created and then it is attached to the
+BMH on its next reboot/power on.
+
+The `DataImage` `Status` will cache the image details including the attachment
+status, which will inform us if the attachment succeeded and can also be used
+for detachment - either when the attachment fails and we retry or when the user
+explicitly requests detachment by deleting the `DataImage` CRD that corresponds to a BMH.
+
+We will introduce a new flag, `NonBootableISOImage` (under `pkg/hardwareutils/bmc`), at the
+webhook level to validate if a given driver supports the non-bootable ISO feature.
+
+### Implementation Details
+
+As part of the design we are assuming that to attach an image to a BMH object, the user
+will create the corresponding `DataImage` CR (with the same `Name` and `Namespace` as the
+`BareMetalHost` object). The BMH Controller will add the ownerReference for the corresponding BMH
+(example : [HostFirmwareSettings](https://github.com/metal3-io/baremetal-operator/blob/9ce684e6462a6ab55ff650cb6c11f9ba1ffb395d/docs/api.md?plain=1#L664)) and it will also add a finalizer to the `DataImage` after its attached.
+Then, on the next reboot/power-on of the BMH, the `DataImage` will be attached and the `DataImage`
+`Status` will be updated to reflect the status of the attachment, including which image was attached.
+In case of detachment of the ISO(`DataImage`), similar process will be followed where the actual
+detachment will happen on the next reboot/power-on and the `DataImage` Status will reflect the status
+of detachment, example : number of failures. The finalizer will also be removed once the
+detachment succeeds.
+Here is an example of what the `DataImage` `Status` might look like:
+
+```yaml
+status:
+  lastReconciliation: "2024-01-01T12:00:00Z"
+  error:
+    count: 0
+    message: ""
+  attachedImage:
+    url: "http://example.com/images/dataimage.iso"
+```
+
+Since the attachment of the ISO happens when a BMH reconcile is triggered as a result of BMH
+reboot/power on, so handling the attachment/detachment of dataImage when the host reaches steady state (refer [`actionManageSteadyState` function](https://github.com/metal3-io/baremetal-operator/blob/9ce684e6462a6ab55ff650cb6c11f9ba1ffb395d/controllers/metal3.io/baremetalhost_controller.go#L1414) )
+makes sense since we only reach this state when a host has been either provisioned or externallyProvisioned, and
+any power state changes (including reboot via rebootAnnotation) are also handled after the host reaches this state
+( refer [`manageHostPower` function](https://github.com/metal3-io/baremetal-operator/blob/1bb45eef449c942711b1c0937ecff2b10a326eb3/controllers/metal3.io/baremetalhost_controller.go#L222) )
```

Comment:
So what is the expected flow?

1. User provisions a node (BMH is `Provisioned`)
2. User creates DataImage
3. User forces a BMH reboot

Or is the DataImage supposed to exist when the BMH moves to the Provisioned or ExternallyProvisioned state?

Alternatively will this work?

1. Create BMH as `ExternallyProvisioned` and `online: false`
2. Create `DataImage`
3. Set BMH `online: true`

Reply:
I think both these workflows should work. Basically, if the BMH has reached the `Provisoned`/`ExternallyProvisioned` state and a valid `DataImage` exists, then on the next reboot the attachment will happen.